### ISSUE TITLE
You can't wrap async tests within an async describe.

### ISSUE THREAD

**Issue by deathlynocturnal** [nodejs/node-core-test#49]
I have switched to using `node:test` for my testing library, and so far it's been perfect, but only noticed that I could do sub-tests which is perfect for a given use case but found some issues with `async` describe.

Node Version:
```
v20.3.1
v20.5.1 (tested with this version too, see comments)
```

Worth noting I am using `--loader tsx`

Command below:
```
node --loader tsx --test ./src/**/*.spec.ts
```

Take this code:

```ts
import test, { describe } from 'node:test';
import assert from 'node:assert';

describe('parent', () => {
	test('child #1', async t => {
		await t.test('grandchild #1.1', async () => {
			assert(true);
		});
	});

	test('child #2', async t => {
		await t.test('grandchild #2.1', async () => {
			assert(true);
		});
	});

	test('child #3', async t => {
		await t.test('grandchild #3.1', async () => {
			assert(true);
		});
	});
});
```

Output is correct:
```
▶ parent
  ▶ child #1
    ✔ grandchild #1.1 (0.414883ms)
  ▶ child #1 (1.356085ms)

  ▶ child #2
    ✔ grandchild #2.1 (0.128633ms)
  ▶ child #2 (0.315027ms)

  ▶ child #3
    ✔ grandchild #3.1 (0.12717ms)
  ▶ child #3 (0.305297ms)

▶ parent (2.797417ms)

ℹ tests 6
ℹ suites 1
ℹ pass 6
ℹ fail 0
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 157.79426
```

---

Now take the same code above, but instead describe is going to be async, maybe we want to prep something for all tests within that describe method (which is something I would need for my use case anyway).

```ts
import test, { describe } from 'node:test';
import assert from 'node:assert';

describe('parent', async () => {

	// I may want to initialise an async package here, etc.
	// Then feed it to all children inside.

	await test('child #1', async t => {
		await t.test('grandchild #1.1', async () => {
			assert(true);
		});
	});

	await test('child #2', async t => {
		await t.test('grandchild #2.1', async () => {
			assert(true);
		});
	});

	await test('child #3', async t => {
		await t.test('grandchild #3.1', async () => {
			assert(true);
		});
	});
});
```

This now fails and the following is returned:
```
▶ parent
  ▶ child #1
    ✔ grandchild #1.1 (0.54243ms)
  ▶ child #1 (1.95943ms)

▶ parent (3.032669ms)

✖ child #2 (0.403573ms)
  'test could not be started because its parent finished'

✖ child #3 (0.048818ms)
  'test could not be started because its parent finished'

ℹ tests 4
ℹ suites 1
ℹ pass 2
ℹ fail 2
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 152.380846

✖ failing tests:

✖ child #2 (0.403573ms)
  'test could not be started because its parent finished'

✖ child #3 (0.048818ms)
  'test could not be started because its parent finished'
error Command failed with exit code 1.
```

Looking around, I can see some libraries used to allow describe to be async, and now that is not the case.

I appreciate that this may be closed because "others" don't do it, but at least for my use case, I could see this being super useful, and I can't realistically imagine it being _that_ complex to support (please excuse my naivety).

**Comment by benjamingr** [comment]
You would want to run your setup in a before block which can be async and pass it that way I think? I think async describe makes senes if the tests inside the describe are generated based on async data maybe?

> I appreciate that this may be closed because "others" don't do it, but at least for my use case, 

Not at all (because others don't do it) though I think it'd be good to understand why `before` doesn't work for your case?

**Comment by aduh95** [comment]
I think it was already reported in https://github.com/nodejs/node/issues/48845 and fixed subsequently. Can you try with the latest Node.js version?

**Comment by deathlynocturnal** [comment]
> I think it was already reported in [nodejs/node#48845](https://github.com/nodejs/node/issues/48845) and fixed subsequently. Can you try with the latest Node.js version?

Upgraded to v20.5.1 - which I assume is what you mean.
Results are the same unfortunately.

**Comment by deathlynocturnal** [comment]
> You would want to run your setup in a before block which can be async and pass it that way I think?

Yeah I could definitely do it this way, was hoping to avoid boilerplate really, if I am already writing a describe block to group it, feels logical to put it there.

> I think async describe makes senes if the tests inside the describe are generated based on async data maybe?

Yeah, there are various things that I could utilise an async describe method for, I could of course simply do it inside of the test and cache it for other tests (I believe they run in series; not sure on this one), but again, feels logical to just use the describe to do that.

> > I appreciate that this may be closed because "others" don't do it, but at least for my use case,
>
> Not at all (because others don't do it) though I think it'd be good to understand why `before` doesn't work for your case?

Not sure what you mean by `before`? Could you clarify?

**Comment by benjamingr** [comment]
@dannysmc95 you would presumably want v20.6 +

**Comment by benjamingr** [comment]
Oh and
```
import test, { describe } from 'node:test';
import assert from 'node:assert';

describe('parent', async () => {

	let myAsyncPackage;
        before(async () => myAsyncPackage = await initialize());
        // ... rest of code
});
```

**Comment by deathlynocturnal** [comment]
> @dannysmc95 you would presumably want v20.6 +

Ahh annoying, looks like the nodesource apt package I use doesn't have it listed, so I can only update to v20.5.1 as the latest available version, time to play the waiting game.

**Comment by deathlynocturnal** [comment]
> Oh and
> 
> ```
> import test, { describe } from 'node:test';
> import assert from 'node:assert';
> 
> describe('parent', async () => {
> 
> 	let myAsyncPackage;
>         before(async () => myAsyncPackage = await initialize());
>         // ... rest of code
> });
> ```

Interesting, I have not seen this before, but looks pretty interesting if that's a work around, I am not at the point I actually need it yet, still trying it out, but will come back to this when I get there and see if it solves it.

Still boilerplate in my opinion, but I understand there may be some restrictions I don't fully appreciate here.

**Comment by deathlynocturnal** [comment]
Closing this, as the issue is fixed in v20.6.1
